WordPress/gutenberg
# Style Engine: add typography and color to backend

Tracking Issue: https://github.com/WordPress/gutenberg/issues/38167

## What?

An alternative to

- https://github.com/WordPress/gutenberg/pull/40082

that takes into account classname generation from slugs, as well as required classnames such as `has-text-color`, which is applied to blocks even with a custom text color.


## Why?
We need to cater for block support classnames that both:

- require slug substitution
- should display regardless of whether there is preset or inline CSS

## How?
By accepting an array for the block style value that contains metadata.

Example usage:

```php
$block_styles = array(
    'color' => array(
        'text'       => '#ccc',
        'background' => 'var:preset|background-color|burned-toast',
    ),
    'typography' => array(
        'fontSize' =>'var:preset|font-size|biiiiiig',

    ),
    'spacing' => array(
        'margin' = array(
            'top'    => '12rem',
            'left'   => '2vh',
            'bottom' => '2px',
            'right'  => '10em',
        ),
    ),
);


$styles = gutenberg_get_style_engine()->generate( block_styles );

/*
$styles outputs:

array(
    'css' => 'color: #ccc; margin-top: 12rem; margin-left: 2vh; margin-bottom: 2px; margin-right: 10em;',
    'classnames' => 'has-text-color has-background has-burned-toast-background-color has-biiiiiig-font-size',
)
*/
```

## Testing Instructions

For dynamic blocks (Site Title, Navigation, Post Title, Post Date for example), check that the color, spacing and typography block supports CSS and classnames are output correctly on the frontend.

<details>
<summary>Some example block code (using TwentyTwentyTwo theme presets)</summary>

```html













```

</details>

Run `npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/packages/style-engine/phpunit/class-wp-style-engine-test.php`
